util/CircularBuffer.h

```diff
@@ -0,0 +1,198 @@
+/* mbed Microcontroller Library
```

Comment:
Yes, there is crucial differences with the one in mbed os: 
* Synchronization is not present in the class as in our case it is a too big overhead to lock and unlock interrupt whenever a character is read or written in the buffer. In our case we just need to lock for the read operation as the write operation happens in interrupt and cannot be interrupted by the reader.
* The API allows application to push and pop chucks of data; not just a single byte. It would be neat to backport that optimization to mbed-os.